apps/shared-app/src/client/product.tsx

```diff
@@ -15,10 +15,30 @@ export interface ProductProps {
 }
 
 export function Product({buy}: ProductProps): JSX.Element {
-  const [result, formAction] = ReactDOM.useFormState(buy, undefined);
+  const [formState, formAction] = ReactDOM.useFormState(buy, undefined);
+
+  const [result, setOptimisticResult] = React.useOptimistic<
+    BuyResult | undefined,
+    number
+  >(formState, (prevResult, quantity) => ({
+    status: `success`,
+    quantity,
+    totalQuantityInSession: prevResult?.totalQuantityInSession ?? 0 + quantity,
+  }));
 
   return (
-    <form action={formAction}>
+    <form
+      action={formAction}
+      onSubmit={(event) => {
+        event.preventDefault();
+        const formData = new FormData(event.currentTarget);
+
+        React.startTransition(() => {
+          setOptimisticResult(parseInt(formData.get(`quantity`) as string, 10));
+          formAction(formData);
```

Comment:
You should `await` this 😄

Reply:
This can't be awaited, a formAction returns `void`, not `Promise<void>`.

Reply:
Ah good point!

Reply:
> You need to use the hook version so that this transition is attached to the component, and then await the action. Otherwise things like errors from the action won't be surfaced.

How can I verify this statement? Using `React.startTransition`, at least the component stack looks fine when forcing a network error:
<img width="610" alt="Screenshot 2024-01-12 at 20 00 54" src="https://github.com/unstubbable/mfng/assets/761683/657411fc-fa39-42bf-abf3-597b91c3705c">

Reply:
@tom-sherman Sounds to me like the error behavior you've described here, which I couldn't reproduce, is addressed in [facebook/react#28111](https://github.com/facebook/react/pull/28111). Am I interpreting this correctly?